src/qibosoq/qick_programs.py

```diff
@@ -33,62 +29,62 @@ def __init__(self, soc: QickSoc, qpcfg: QickProgramConfig, sequence: PulseSequen
             * wait_initialize is defined explicitly
             * super.__init__ (this will init AveragerProgram or RAveragerProgram)
         """
-
         self.soc = soc
         self.soccfg = soc  # this is used by qick
 
         self.sequence = sequence
         self.qubits = qubits
 
         # general settings
-        self.max_gain = qpcfg.max_gain
         self.adc_trig_offset = qpcfg.adc_trig_offset
-        self.max_sampling_rate = qpcfg.sampling_rate
         self.reps = qpcfg.reps
 
-        self.relax_delay = self.us2cycles(qpcfg.repetition_duration * NS_TO_US)
+        # mux settings
+        self.is_mux = qibosoq_cfg.IS_MULTIPLEXED
+        self.readouts_per_experiment = None
+
+        self.relax_delay = self.us2cycles(qpcfg.repetition_duration)
         self.syncdelay = self.us2cycles(0)
         self.wait_initialize = self.us2cycles(2.0)
 
         self.pulses_registered = False
         self.registered_waveform = []
 
+        if self.is_mux:
+            self.multi_ro_pulses = self.create_mux_ro_dict()
+            self.readouts_per_experiment = len(self.multi_ro_pulses)
+
         # pylint: disable-next=too-many-function-args
         super().__init__(soc, asdict(qpcfg))
 
-    def declare_nqz_zones(self, sequence: PulseSequence):
+    def declare_nqz_zones(self, sequence: List[Pulse]):
         """Declare nqz zone (1-2) for a given PulseSequence
 
         Args:
             sequence (PulseSequence): sequence of pulses to consider
         """
-
         ch_already_declared = []
         for pulse in sequence:
-            if pulse.type is PulseType.DRIVE:
-                gen_ch = self.qubits[pulse.qubit].drive.ports[0][1]
-            elif pulse.type is PulseType.READOUT:
-                gen_ch = self.qubits[pulse.qubit].readout.ports[0][1]
-
             freq = pulse.frequency
+            gen_ch = pulse.dac
 
             if gen_ch not in ch_already_declared:
                 ch_already_declared.append(gen_ch)
-                zone = 1 if freq < self.max_sampling_rate / 2 else 2
+                sampling_rate = self.soccfg["gens"][gen_ch]["fs"]
+                zone = 1 if freq < sampling_rate / 2 else 2
                 self.declare_gen(gen_ch, nqz=zone)
 
     def declare_readout_freq(self):
         """Declare ADCs downconversion frequencies"""
-
         adc_ch_already_declared = []
-        for readout_pulse in self.sequence.ro_pulses:
-            adc_ch = self.qubits[readout_pulse.qubit].feedback.ports[0][1]
-            ro_ch = self.qubits[readout_pulse.qubit].readout.ports[0][1]
+        for readout_pulse in [pulse for pulse in self.sequence if pulse.type == "readout"]:
```

Comment:
```suggestion
        for readout_pulse in (pulse for pulse in self.sequence if pulse.type == "readout"):
```

No practical difference, but just to let you know: this is the optimal place for a generator, because in this way the iteration will only happen once (during the `for` loop), just filtering the iterator.
Instead, if you use list comprehension, it would be iterated twice: first to construct the list (that is only used once, and then thrashed), and then for the `for` loop.

Since you're not modifying the elements, in this case another option would be:
```suggestion
        for readout_pulse in filter(lambda p: p.type == "readout", self.sequence):
```

In this case they are both fine by me. In general, comprehension is more powerful, because it's the combination of `map()` and `filter()`, with an elegant syntax. But since you only need `filter()`, here they are both fine.